---
title: "Gene variant level analysis"
author: "digital ECMT cancer trial matching tool"
date: "05/03/2021"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = T)
require(dplyr)
require(formattable)
require(tidyr)
require(ggraph)
require(igraph)
require(networkD3)
require(tidygraph)
require(ggraph)
require(visNetwork)
```



```{r download data}
## download variant summaries from clinvar
download.file(url="https://ftp.ncbi.nlm.nih.gov/pub/clinvar/tab_delimited/variant_summary.txt.gz",destfile = "variant_summary.txt.gz")
clinvar <- read.table(gzfile("variant_summary.txt.gz"), sep="\t", quote="", fill = TRUE, stringsAsFactors = FALSE) 
# V1 contains values for Allele ID
# V2 gives nature (e.g. fusion, single nucleotide variant etc)
# V7 contains values for clinical significance (inc whether pathogenic or not)
# V25 gives a measure of confidence in assertion
clinvar <- unique(dplyr::select(clinvar, "allele.id" = "V1", "nature"="V2", "significance"="V7", "confidence"="V25"))

## variation_allele contains mappings from clinvar variation ID (clinvar ID) and allele ID
download.file(url="https://ftp.ncbi.nlm.nih.gov/pub/clinvar/tab_delimited/variation_allele.txt.gz",destfile = "variation_allele.txt.gz")
variationAllele <- read.table(gzfile("variation_allele.txt.gz"), sep="\t", quote="", fill = TRUE, stringsAsFactors = FALSE) 
# V1 contains values for clinvar ID
# V3 contains values for Allele ID
variationAllele <- unique(dplyr::select(variationAllele, "clinvar.id"="V1", "allele.id"="V3"))

## join tables
clinvar <- merge(x=clinvar, by.x="allele.id", y=variationAllele, by.y="allele.id")




# variants contains mappings from variant_id to clinvar_ids
variants <- read.csv(file = "https://civicdb.org/downloads/nightly/nightly-VariantSummaries.tsv", sep = "\t", quote = "", stringsAsFactors = FALSE)
# variants$gene contains gene name (Entrez symbol)
# variants$variant contains amino acid change
# variants$summary contains a useful summary of evidence for each variant
# variants$clinvar_ids contains (comma-separated) clinvar IDs that can be mapped to the clinvar table to get info whether pathogenic or not... 

## drop unwanted columns
variants <- unique(dplyr::select(variants, variant_id, gene, variant, summary, variant_groups,civic_variant_evidence_score,clinvar_ids ))

## split and unnest the clinvar_ids column
variants$clinvar_ids <- strsplit(as.character(variants$clinvar_ids), split = ",")
variants <- unnest(data = variants, clinvar_ids)


## join significance from clinvar
variants <- unique(merge(x=variants, by.x="clinvar_ids", all.x=TRUE, y=clinvar, by.y="clinvar.id"))


## download evidence from civicDB
# evidence contains mappings from variant_id to gene (gene name) and variant (amino acid change) to drugs etc
evidence <- read.csv(file = "https://civicdb.org/downloads/nightly/nightly-ClinicalEvidenceSummaries.tsv", sep = "\t", quote="", stringsAsFactors = FALSE)
# evidence contains more detailed data re: drugs, sensitivity etc
# evidence$evidence_level contains details of nature of evidence: 
# A = validated
# B = clinical
# C = case study
# D = preclinical


## drop unwanted columns
evidence <- unique(dplyr::select(evidence, variant_id, gene, variant,disease, drugs, drug_interaction_type, evidence_type, evidence_direction,evidence_level,clinical_significance,evidence_statement, source_type))

evidence <- as.data.frame(evidence)

## merge with variants
#evidence <- unique(merge(x=evidence, by.x="variant_id", all.x=TRUE, y=unique(dplyr::select(variants, variant_id, clinvar_ids, nature, significance, confidence)),by.y="variant_id" ))
```
  
### **Example data from civicDB variants table**  
  
#### **For specific variant**  
  
```{r specific variant}

formattable(filter(variants, gene=="NRAS" & variant=="Q61R"))


```


#### **For amino acid position**  
  
```{r amino acid position}

formattable(filter(variants, gene=="NRAS" & variant=="Q61"))


```
  
#### **For any mutation**  
  
```{r any mutation}  


formattable(filter(variants, gene=="NRAS" & variant=="Mutation"))

```


### **Example data from civicDB evidence table**  
  
#### **For specific variant**  
  
```{r evidence for specific variant}

formattable(filter(evidence, gene=="NRAS" & variant=="Q61R"))


```


#### **For amino acid position**  
  
```{r evidence for amino acid position}

formattable(filter(evidence, gene=="NRAS" & variant=="Q61"))


```
  
#### **For any mutation**  
  
```{r evidence for any mutation}  


formattable(filter(evidence, gene=="NRAS" & variant=="Mutation"))

```
  
### **What proportion of drugs listed in civicDB evidence are found in NCI thesaurus?**  
  

```{r download and create NCIthesaurus table}

## specify URL for NCI thesaurus - this should always be the most recent? 
NCItURL <- "https://evs.nci.nih.gov/ftp1/NCI_Thesaurus/Thesaurus.FLAT.zip"

destFlatFilename <- "NCIt_FLAT.zip"
download.file(url=NCItURL,destfile = destFlatFilename)
unzip(zipfile = paste0(getwd(),"/",destFlatFilename))

NCIt <- read.table("Thesaurus.txt",header = FALSE, sep = "\t", comment.char = "", fill = TRUE, stringsAsFactors = FALSE, quote = "")
names(NCIt) <- c("ID","URL","ParentID","Synonyms","Description","PreferredTerm","Type","Class")

## a drug may have more than one class, so need to multiply rows
## split and unnest the Class column of NCIt
NCIt$Class <- strsplit(NCIt$Class, split = "\\|")
NCIt <- unnest(data = NCIt, Class)

## subset to retain only relevant classes
NCItPharmacologics <- NCIt[which(NCIt$Class %in% c("Pharmacologic Substance","Amino Acid, Peptide, or Protein","Immunologic Factor")), ]

## one drug can have more than one parent, so need to split into multiple rows...
## split ParentID column on pipe symbol
NCItPharmacologics$ParentID <- strsplit(NCItPharmacologics$ParentID, split = "\\|")
## unnest the ID column to multiply rows
NCItPharmacologics <- unnest(data=NCItPharmacologics,ParentID)

## join parent synonyms
## first, get the synonyms and IDs for parents
NCItParents <- NCIt[which(NCIt$ID %in% NCItPharmacologics$ParentID), which(names(NCIt) %in% c("ID", "Synonyms"))]

## each parent term may have more than one synonym
## for simplicity, we will retain only the first synonym for each parent term
NCItParents$Synonyms <- gsub("\\|.*","",NCItParents$Synonyms)

## merge parents on entity ID = parent ID
NCItPharmacologics <- merge(x=NCItPharmacologics,y=NCItParents,by.x="ParentID",by.y="ID",all.x=TRUE)
## reorder and rename
NCItPharmacologics <- dplyr::select(NCItPharmacologics,ID,Class,PreferredTerm,Synonyms="Synonyms.x",Description,ParentTerm="Synonyms.y",ParentID)

## split the Synonyms column on pipe symbol
NCItPharmacologics$Synonyms <- strsplit(NCItPharmacologics$Synonyms, split="\\|")
## unnest the Synonyms column to multiply rows
NCItPharmacologics <- unnest(data = NCItPharmacologics,Synonyms)

## add a column with lower case drug synonyms for joining
NCItPharmacologics$SynonymsLower <- tolower(NCItPharmacologics$Synonyms)

## convert from tibble to data frame
NCIthesaurus <- as.data.frame(NCItPharmacologics)
NCItPharmacologics <- as.data.frame(NCItPharmacologics)


## add a column to indicate date downloaded
NCIthesaurus$downloaded <- Sys.Date()

## write to comma separated file
#write.csv(x=NCIthesaurus,file="NCIthesaurus.csv",row.names=FALSE)


## check written correctly
#dim(read.csv("NCIthesaurus.csv", header = TRUE, stringsAsFactors = FALSE))
#names(read.csv("NCIthesaurus.csv", header = TRUE, stringsAsFactors = FALSE))

```
   
```{r unnest evidence drugs, echo=TRUE}
## split and unnest the clinvar_ids column
evidence$drugs <- strsplit(evidence$drugs, split=",")
evidence <- unnest(data=evidence, drugs)
# variants$clinvar_ids <- strsplit(as.character(variants$clinvar_ids), split = ",")
# variants <- unnest(data = variants, clinvar_ids)

## how many drugs listed?
drugsInEvidence <- unique(tolower(evidence$drugs))
length(drugsInEvidence)


## how many drug synonyms listed in NCI thesaurus?
drugsInNCIt <- unique(NCItPharmacologics$SynonymsLower)
length(drugsInNCIt)


## how many drugs are in evidence$drugs but not NCItPharmacologics$synonyms?
drugsNotInNCIt <- setdiff(drugsInEvidence, drugsInNCIt)
length(drugsNotInNCIt)

## what proportion of drugs in evidence$drugs are not in NCItPharmacologics?
length(drugsNotInNCIt)/length(drugsInEvidence)

## how many "drugs" not in NCIt are actually mechanisms, and which are in NCIt parent terms?
NCItParentTerms <- unique(tolower(NCItPharmacologics$ParentTerm))

setdiff(drugsNotInNCIt,NCItParentTerms  )

```

### **Graph visualisation**  
  
```{r load tables from output of trialMatchDataRefresh}  
## read cancerStudies table into memory
# contains study details, mapped to a controlled set of cancer types in the "TARGET.condition"  column
# names are "locations, nct_id, brief_title, overall_status, condition, interventions, postcode, site_status, investigators, contacts, Refresh.date, matching.condition, TARGET.condition, Link, lat, long"
# no columns are aggregated
cancerStudies <- unique(read.table(file = "cancerStudies.tsv", header = TRUE, stringsAsFactors = FALSE))

## filter and retain only studies with overall status of "Recruiting"
cancerStudies <- dplyr::filter(cancerStudies, overall_status == "Recruiting")

## filter out sites with status of not yet recruiting
cancerStudies <- dplyr::filter(cancerStudies, site_status == "Recruiting")



targetMatches <- unique(read.table(file = "targetMatches.tsv", header = TRUE, stringsAsFactors = FALSE))


manualNLPmatches <- unique(read.table(file = "manualNLPmatches.tsv", header = TRUE, stringsAsFactors = FALSE))

pathwayMatches <- unique(read.table(file = "pathwayMatches.tsv", header = TRUE, stringsAsFactors = FALSE))

```




```{r create node and edge lists, echo=TRUE}
## gene to drug edges from targetMatches
gene.target.edges <- unique(dplyr::select(targetMatches, "from" = "Target", "to"="Interventions"))
gene.target.edges$to <- tolower(gene.target.edges$to)

gene.target.edges <- gene.target.edges[complete.cases(gene.target.edges), ]
gene.target.edges$title <- "targeted.by"
## get gene and drug nodes
gene.nodes <- unique(dplyr::select(targetMatches, "id" = "Target"))
drug.nodes <- unique(dplyr::select(targetMatches, "id"="Interventions"))



## preview last few rows of tables
formattable(tail(gene.target.edges))
formattable(tail(gene.nodes))
formattable(tail(drug.nodes))



## gene to study edges from manualNLPmatches
gene.study.edges <- unique(dplyr::select(manualNLPmatches, "from" = "matching.symbol", "to"="nct_id", "title"="criteria"))
gene.nodes <- rbind(gene.nodes, unique(dplyr::select(manualNLPmatches, "id" = "matching.symbol")))
study.nodes <- unique(dplyr::select(manualNLPmatches, "id"="nct_id"))

formattable(tail(gene.study.edges))
formattable(tail(gene.nodes))
formattable(tail(study.nodes))

## gene to downstream gene edges from pathwayMatches
gene.downstream.edges <- unique(dplyr::select(pathwayMatches, "from" = "network.element.gene.names", "to"="downstream.genes", "title"="downstream"))
gene.nodes <- rbind(gene.nodes, unique(dplyr::select(pathwayMatches, "id" = "network.element.gene.names")), unique(dplyr::select(pathwayMatches, "id" = "downstream.genes")))
formattable(tail(gene.downstream.edges))
formattable(tail(gene.nodes))




## interventions to studies from cancerStudies
drug.study.edges <- unique(dplyr::select(cancerStudies, "from" = "interventions", "to"="nct_id"))
drug.study.edges$from <- tolower(drug.study.edges$from)
drug.study.edges$title <- "included.in"
drug.nodes <- rbind(drug.nodes, unique(dplyr::select(cancerStudies, "id" = "interventions")))
study.nodes <- rbind(study.nodes, unique(dplyr::select(cancerStudies, "id" = "nct_id")))

formattable(tail(drug.study.edges))
formattable(tail(drug.nodes))
formattable(tail(study.nodes))



## variant to drug edges from evidence
# subset evidence
evidence.edges <- unique(dplyr::select(evidence, gene, variant, drugs, evidence_direction, clinical_significance, evidence_statement))
# concatenate evidence direction and clinical_significance
evidence.edges$evidence_statement <- paste(toupper(evidence.edges$evidence_direction), toupper(evidence.edges$clinical_significance), evidence.edges$evidence_statement, sep="; ")

# create gene_variant column
evidence.edges$gene_variant <- paste(evidence.edges$gene, evidence.edges$variant, sep="_")
# create edge lists
variant.drug.edges <- unique(dplyr::select(evidence.edges, "from"="gene_variant", "to"="drugs", "title"="evidence_statement"))
variant.drug.edges$to <- tolower(variant.drug.edges$to)

variant.nodes <- unique(dplyr::select(evidence.edges, "id"="gene_variant"))
drug.nodes <- rbind(drug.nodes, unique(dplyr::select(evidence.edges, "id" = "drugs")))

formattable(tail(variant.drug.edges))
formattable(tail(variant.nodes))
formattable(tail(drug.nodes))





## variant to gene from variants
# create a copy 
variant.edges <- unique(dplyr::select(variants, gene, variant, significance))
variant.edges$gene_variant <- paste(variant.edges$gene, variant.edges$variant, sep="_")
variant.gene.edges <- unique(dplyr::select(variant.edges, "from"="gene_variant", "to"="gene"))
variant.gene.edges$title <- NA
variant.nodes <- rbind(variant.nodes, unique(dplyr::select(variant.edges, "id" = "gene_variant")))
gene.nodes <- rbind(gene.nodes, unique(dplyr::select(variants, "id" = "gene")))

formattable(tail(variant.gene.edges))
formattable(tail(variant.nodes))
formattable(tail(gene.nodes))

variant.nodes$group <- "variant"
gene.nodes$group <- "gene"
drug.nodes$group <- "drug"
study.nodes$group <- "study"

variant.nodes$color <- "red"
gene.nodes$color <- "blue"
drug.nodes$color <- "green"
study.nodes$color <- "black"

## convert all drug names to lower case to remove redundancy
drug.nodes$id <- tolower(drug.nodes$id)

## combine into a single edge list
edgeList <- unique(rbind(gene.target.edges, gene.study.edges, gene.downstream.edges, drug.study.edges, variant.drug.edges, variant.gene.edges))

nodeList <- unique(rbind(variant.nodes, gene.nodes, drug.nodes, study.nodes))     
## take out any redundancy
# nodeList <- as.data.frame(nodeList %>% 
#                   group_by(id) %>%
#                   summarise(
#                       group=paste(unique(group), collapse = "; ")
#                   ))
nodeList <- nodeList[complete.cases(nodeList), ]
edgeList <- edgeList[complete.cases(edgeList), ]
```


```{r create node lists, echo=T}
# ## gene nodes 
# gene.nodes <- data.frame(id=unique(c(targetMatches$Target, gene.downstream.edges$from, gene.downstream.edges$to, evidence$edges$gene, manualNLPmatches$matching.symbol, variant.gene.edges$to)))
# gene.nodes$class <- "gene"
# gene.nodes$popup <- NA
# formattable(head(gene.nodes))
# ## variant nodes from variant.edges
# # need to account for redundanc in significance column
# variant.nodes <- as.data.frame(variant.edges %>%
#                   group_by(id=gene_variant) %>%
#                   summarise(
#                             popup = paste(unique(significance), collapse = "; "), 
#                             class = "variant"))
# formattable(head(variant.nodes))
# ## drug nodes
# drug.nodes <- data.frame(id=unique(cancerStudies$interventions))
# drug.nodes$popup <- NA
# drug.nodes$class <- "drug"
# formattable(head(drug.nodes))
# ## study nodes <- 
# study.nodes <- unique(dplyr::select(cancerStudies, "id" = "nct_id", "popup"="brief_title"))
# study.nodes$class <- "study"
# formattable(head(study.nodes))
## combine into a single node list
#nodeList <- unique(rbind(gene.nodes, variant.nodes, drug.nodes, study.nodes))
```






```{r create graphs, echo=T}
basic.igraph <- graph_from_data_frame(d=edgeList, directed=T, vertices = nodeList)
class(basic.igraph)

## subset for nodes downstream of NRAS
# first, get a list of all nodes reachable from (i.e downstream of) NRAS
downstreamNodes <- subcomponent(basic.igraph, "NRAS", mode = c("out"))
downstreamNodes
# second, create a subgraph of basic.graph, containing only the specified vertices and all the edges among them
downstreamGraph <- induced_subgraph(graph=basic.igraph, vids = downstreamNodes, impl = "auto")
# third, plot downstream subgraph
plot(downstreamGraph, layout = layout_with_graphopt, edge.arrow.size = 0.2)
# layout as tree
# create layout object
l <- layout_as_tree(downstreamGraph, root=which(V(downstreamGraph)$name == "NRAS"),mode="out")


plot(downstreamGraph, layout=l)

## hide edge titles
plot(downstreamGraph, layout=l, edge.label = NA)



```


see https://www.jessesadler.com/post/network-analysis-with-r/ for more info on graph layouts and options,   
also https://kateto.net/network-visualization  
  
  



```{r tidygraph visualisations, warning=F}



## convert to tidygraph
downstreamGraph_tidy <- as_tbl_graph(downstreamGraph)
## can now configure similar to ggplot
ggraph(downstreamGraph_tidy, l) + 
    geom_node_point(aes(size=2, alpha=0.5, color=V(downstreamGraph)$class)) +
    geom_edge_link(aes(width = 1), alpha = 0.8,show.legend = FALSE) + 
    scale_edge_width(range = c(0.2, 2)) +
    geom_node_text(label=V(downstreamGraph)$name, repel = TRUE) +
    guides(alpha = FALSE, size = FALSE) +
    theme_graph()





```




```{r visNetwork visualisations}
#library(visNetwork)
visIgraph(downstreamGraph)




## size nodes based on degree centrality (more links -> larger node) We will also change the width of the edges based on their weight.
# Compute node degrees (#links) and use that to set node size:
deg <- degree(downstreamGraph, mode="all")
V(downstreamGraph)$size <- deg*3


visIgraph(downstreamGraph)

visIgraph(downstreamGraph) %>%
    visIgraphLayout(layout = "layout_with_fr") %>%
    visEdges(arrows = "middle")



## layout as tree, with root nodes at the top
visIgraph(downstreamGraph) %>%
    visHierarchicalLayout(sortMethod = "directed", direction = "UD")%>%
    visEdges(arrows = "middle")


## selectable nodes
visIgraph(downstreamGraph) %>%
    visHierarchicalLayout(sortMethod = "directed", direction = "UD")%>%
    visOptions(highlightNearest = TRUE, nodesIdSelection = TRUE) %>%
    visEdges(arrows = "middle")

## show labels on hover
# visIgraph(downstreamGraph) %>%
#   visHierarchicalLayout(sortMethod = "directed", direction = "UD")%>%
#   visEdges(arrows = "middle") %>%
#   visInteraction(hover = T) %>%
#   visEvents(hoverEdge = "function(e){
#     this.body.data.edges.update({id: e.edge, font: {size : 14}});
#   }", hoverNode = "function(f){
#     this.body.data.nodes.update({id: f.node, font: {size : 14}});
#   }") %>%
#   visEvents(blurEdge = "function(e){
#     this.body.data.edges.update({id: e.edge, font: {size : 0}});
#   }", blurNode = "function(f){
#     this.body.data.nodes.update({id: f.node, font: {size : 0}});
#   }")



# visIgraph(downstreamGraph) %>%
#     visHierarchicalLayout(sortMethod = "directed", direction = "UD")%>%
#     visInteraction(hover = T) %>%
#     visEvents(hoverEdge = "function(e){
#     this.body.data.edges.update({id: e.edge, font: {size : 14}});
#   }") %>%
#     visEvents(blurEdge = "function(e){
#     this.body.data.edges.update({id: e.edge, font: {size : 0}});
#   }") %>%
#     visEvents(hoverNode = "function(e){
#     this.body.data.nodes.update({id: e.node, font: {size : 14}});
#   }") %>%
#     visEvents(blurNode = "function(e){
#     this.body.data.nodes.update({id: e.node, font: {size : 0}});
#   }")



## add selectable nodes 
# visIgraph(downstreamGraph) %>%
#     visHierarchicalLayout(sortMethod = "directed", direction = "UD")%>%
#     visOptions(highlightNearest = TRUE, nodesIdSelection = TRUE) %>%
#     visInteraction(hover = T) %>%
#     visEvents(hoverEdge = "function(e){
#     this.body.data.edges.update({id: e.edge, font: {size : 14}});
#   }") %>%
#     visEvents(blurEdge = "function(e){
#     this.body.data.edges.update({id: e.edge, font: {size : 0}});
#   }") %>%
#     visEvents(hoverNode = "function(e){
#     this.body.data.nodes.update({id: e.node, font: {size : 14}});
#   }") %>%
#     visEvents(blurNode = "function(e){
#     this.body.data.nodes.update({id: e.node, font: {size : 0}});
#   }")

## colour nodes by group
# visIgraph(downstreamGraph) %>%
#     visHierarchicalLayout(sortMethod = "directed", direction = "UD")%>%
#     visOptions(highlightNearest = TRUE, nodesIdSelection = TRUE) %>%
#     # visGroups(groupname = "gene", color = "red", shape = "square", 
#     #           shadow = list(enabled = TRUE)) %>%
#     visInteraction(hover = T) %>%
#     visEvents(hoverEdge = "function(e){
#     this.body.data.edges.update({id: e.edge, font: {size : 14}});
#   }") %>%
#     visEvents(blurEdge = "function(e){
#     this.body.data.edges.update({id: e.edge, font: {size : 0}});
#   }") %>%
#     visEvents(hoverNode = "function(e){
#     this.body.data.nodes.update({id: e.node, font: {size : 14}});
#   }") %>%
#     visEvents(blurNode = "function(e){
#     this.body.data.nodes.update({id: e.node, font: {size : 0}});
#   }")


## probably better to use visNodes(title=) 
# String or Element. Default to undefined. Title to be displayed when the user hovers over the node. The title can be an HTML element or a string containing plain text or HTML. 
# for details on how to implement, see https://stackoverflow.com/questions/48914686/remove-node-and-edge-label-from-visnetwork-visualiztion-and-display-info-when-cl 

```


```{r networkD3 visualisations}
## networkD3 graph
# simple.graph <- simpleNetwork(edgeList, height="100px", width="100px")
# class(simple.graph)




```

